### PR TITLE
Doc 4286: RPC permission templates 

### PR DESCRIPTION
**Pull request by Niamh25**
New RPC permission template section.

### CODE REVIEW

**Comment by vkolomeyko** [comment]
Note preview site available at: https://niamh.preview.docs.r3.com/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.html

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    parent: corda-5-cli-reference
+section_menu: corda-5-beta
+title: "RPC permission templates"
+---
+
+## Create Common RPC Permission Templates
+
+Currently, when the C5 Cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. The idea is to create fine-grained roles for specific actions such as:
+
+• A dedicated role which can create users, roles, permissions and drive all the associations between them;
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:16]: Probably needs an extra empty line or else it is rendered like so:
image


```suggestion
• A dedicated role which can create users, roles, permissions and drive all the associations between them;

```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+• A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload);
+• A dedicated role which allows flows to run on this vNode.
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:18]:
```suggestion
• A dedicated role which allows flows to run on a previously created vNode.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+{{< note >}}
+These roles and permissions enable certain common operations. Creation of users and associating such users to the roles should be still done manually by an admin user.
+{{< /note >}}
+
+UserAdminRole- Creates roles, permissions, and controls all associations between the user, roles, and permissions. Create this role at cluster bootstrap by the admin user. There is no direct write into DB like previously done with the admin user.
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:24]:
```suggestion
`UserAdminRole` - Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+VNodeCreatorRole- Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:26]:
```suggestion
`VNodeCreatorRole` - Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+CordaDeveloperRole- Users can do the following:
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:28]:
```suggestion
`CordaDeveloperRole` - Users can do the following:
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+1. VNode reset
+2. Vnode status update
+
+This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.
+
+FlowExecutorRole- Once the vNode is created, a new CLI command will permit creation of such role for a given vNode.
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:35]:
```suggestion
`FlowExecutorRole` - Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+Being unable to do this creates potentially duplicated permissions.
+{{< /note >}}
+
+## Implement Permissions Check When Starting a Particular Flow
+
+Currently, the check is made to assert whether a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.
+
+These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.
+
+## RPC Permission Templates: Test Plan Draft
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:52]: I wonder if we need to include "Test Plan Draft" into potentially end-user facing document?
Feels like it is more of an internal script which our QA team been following.

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+10. Using Swagger UI, log on as "FlowExecutor" and run a flow ensuring that it successfully completes with expected result returned. When checking results of the flow make sure that vNode ID: 9B02C806787D is used, the calls will be forbidden for any other vNodes.
+11. Execute some negative testing scenarios, e.g.:
+• Ensure that "UserAdmin" cannot upload CPIs, create vNodes or run flows.
+Ensure that "FlowExecutor" and "vNodeCreator" cannot perform any RBAC operations.
+
+## Create RBAC Operation to Create and Assign Multiple Permissions to a Role
+
+A dedicated HTTP RPC call which will permit the following:
+
+• Create multiple permissions
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:91]:
```suggestion
• Create multiple permissions

```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+• Assign them all to the existing roles in a single call
+
+## Allow the vNodeld to Pass when Performing RBAC
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:94]: I wonder if we should delete this section completely?

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md (hunk 1 above)
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:16]:
```suggestion
* A dedicated role which can create users, roles, permissions and drive all the associations between them.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+section_menu: corda-5-beta
+title: "RPC permission templates"
+---
+
+## Create Common RPC Permission Templates
+
+Currently, when the C5 Cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. The idea is to create fine-grained roles for specific actions such as:
+
+• A dedicated role which can create users, roles, permissions and drive all the associations between them;
+• A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload);
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:17]:
```suggestion
* A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload).
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md (hunk 2 above)
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:18]:
```suggestion
* A dedicated role which allows flows to run on this vNode.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+• A dedicated role which allows flows to run on this vNode.
+
+{{< note >}}
+These roles and permissions enable certain common operations. Creation of users and associating such users to the roles should be still done manually by an admin user.
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:21]:
```suggestion
These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md (hunk 3 above)
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:24]: I can't quite follow this an the following lines. Seems like it should be a table or subsections maybe?

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md (hunk 8 above)
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:91]:
```suggestion
* Create multiple permissions
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+9. Use Swagger UI, as "UserAdmin" to create user "FlowExecutor" and assign "FlowExecutorRole-9B02C806787D" to it.
+10. Using Swagger UI, log on as "FlowExecutor" and run a flow ensuring that it successfully completes with expected result returned. When checking results of the flow make sure that vNode ID: 9B02C806787D is used, the calls will be forbidden for any other vNodes.
+11. Execute some negative testing scenarios, e.g.:
+• Ensure that "UserAdmin" cannot upload CPIs, create vNodes or run flows.
+Ensure that "FlowExecutor" and "vNodeCreator" cannot perform any RBAC operations.
+
+## Create RBAC Operation to Create and Assign Multiple Permissions to a Role
+
+A dedicated HTTP RPC call which will permit the following:
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:89]:
```suggestion
A dedicated HTTP RPC call which permits the following:
```

This line needs more info I think?

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+
+• Create multiple permissions
+• Assign them all to the existing roles in a single call
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:92]:
```suggestion
* Assign them all to the existing roles in a single call
```

all what?

**Review by nadinequinr3 — CHANGES_REQUESTED** [review]

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 14]
```diff
@@ -0,0 +1,98 @@
+---
+date: '2023-01-10'
+menu:
+  corda-5-beta:
+    identifier: corda-5-beta-rpc-permissions-templates
+    weight: 6000
+    parent: corda-5-cli-reference
+section_menu: corda-5-beta
+title: "RPC permission templates"
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:9]:
```suggestion
title: "RPC Permission Templates"
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+---
+
+## Create Common RPC Permission Templates
+
+Currently, when the C5 Cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. The idea is to create fine-grained roles for specific actions such as:
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:14]:
```suggestion
By default, when a cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. RPC permission templates enable you to create fine-grained roles for specific actions such as:
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+menu:
+  corda-5-beta:
+    identifier: corda-5-beta-rpc-permissions-templates
+    weight: 6000
+    parent: corda-5-cli-reference
+section_menu: corda-5-beta
+title: "RPC permission templates"
+---
+
+## Create Common RPC Permission Templates
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:12]:
```suggestion

```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+CordaDeveloperRole- Users can do the following:
+
+1. VNode reset
+2. Vnode status update
+
+This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.
+
+FlowExecutorRole- Once the vNode is created, a new CLI command will permit creation of such role for a given vNode.
+
+## Ability to query permissions via HTTP RPC
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:37]:
```suggestion
## Querying Permissions via HTTP RPC
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+After creation role name should be: "FlowExecutorRole-9B02C806787D"
+{{< /note >}}
+
+9. Use Swagger UI, as "UserAdmin" to create user "FlowExecutor" and assign "FlowExecutorRole-9B02C806787D" to it.
+10. Using Swagger UI, log on as "FlowExecutor" and run a flow ensuring that it successfully completes with expected result returned. When checking results of the flow make sure that vNode ID: 9B02C806787D is used, the calls will be forbidden for any other vNodes.
+11. Execute some negative testing scenarios, e.g.:
+• Ensure that "UserAdmin" cannot upload CPIs, create vNodes or run flows.
+Ensure that "FlowExecutor" and "vNodeCreator" cannot perform any RBAC operations.
+
+## Create RBAC Operation to Create and Assign Multiple Permissions to a Role
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:87]: Can we make this heading shorter?

**Review by nadinequinr3 — CHANGES_REQUESTED** [review]

**Review by yarivus — APPROVED** [review]
LGTM

**Comment by yarivus** [comment]
@Niamh25 - please consult @colin-hillis on how to create new branches for PRs with the <author>/<jira-ticket>/<short-desc> format.
Thanks.

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+## Check Permissions When Starting Flows
+
+Currently, Corda checks if a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.
+
+These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.
+
+## RPC Permission Templates
+
+This describes a draft manual QA script that can be executed to independently verify that functionality is working as expected.
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:53]: I still think that we should remove "manual QA script" from this document.

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+After creation role name should be: "FlowExecutorRole-9B02C806787D"
+{{< /note >}}
+
+9. Use Swagger UI, as "UserAdmin" to create user "FlowExecutor" and assign "FlowExecutorRole-9B02C806787D" to it.
+10. Using Swagger UI, log on as "FlowExecutor" and run a flow ensuring that it successfully completes with expected result returned. When checking results of the flow make sure that vNode ID: 9B02C806787D is used, the calls will be forbidden for any other vNodes.
+11. Execute some negative testing scenarios, e.g.:
+• Ensure that "UserAdmin" cannot upload CPIs, create vNodes or run flows.
+Ensure that "FlowExecutor" and "vNodeCreator" cannot perform any RBAC operations.
+
+## Create and Assign Multiple Permissions to a Role
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:86]: This seems unrelated to "RPC Permission Templates" and I propose to remove this section.
OpenAPI call specification along with parameters will be documented along with the rest of the OpenAPI calls.

**Review by vkolomeyko — CHANGES_REQUESTED** [review]

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 21]
```diff
@@ -0,0 +1,76 @@
...
+Being unable to do this creates potentially duplicated permissions.
+{{< /note >}}
+
+## Check Permissions When Starting Flows
+
+Currently, Corda checks if a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.
+
+These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.
+
+## RPC Permission Templates
```
Comment (vkolomeyko) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:45]: @Niamh25 - I still think we should remove this test plan section from the documentation.

Comment (Niamh25) [reply content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md]: @vkolomeyko Apologies, I didn't realise the entire section was to go. It has been removed now!

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+* A dedicated role which allows flows to run on this vNode.
+
+{{< note >}}
+These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
+{{< /note >}}
+
+| Role   | Description |
+| ----------- | ----------- |
+| `UserAdminRole` | Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.       |
+| `VNodeCreatorRole` | Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:26]:
```suggestion
| `VNodeCreatorRole` | Sets all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+section_menu: corda-5-beta
+title: "RPC Permission Templates"
+---
+
+
+By default, when a cluster starts there is a single "admin" account auto-created which can do everything. While this account can be used to perform any action, there is room for error if not used carefully. RPC permission templates enable you to create fine-grained roles for specific actions such as:
+
+* A dedicated role which can create users, roles, permissions and drive all the associations between them.
+* A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload).
+* A dedicated role which allows flows to run on this vNode.
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:17]:
```suggestion
* A dedicated role which allows flows to run on this virtual node.
```
Please use "virtual node" for consistency.

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+
+{{< note >}}
+These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
+{{< /note >}}
+
+| Role   | Description |
+| ----------- | ----------- |
+| `UserAdminRole` | Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.       |
+| `VNodeCreatorRole` | Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
+| `CordaDeveloperRole` | Users can use VNode reset and Vnode status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:27]:
```suggestion
| `CordaDeveloperRole` | Users can use virtual node reset and virtual node status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+| `FlowExecutorRole`|  Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.|
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:28]:
```suggestion
| `FlowExecutorRole`|  Once the virtual node is created, a new CLI command will permit creation of such role for a given virtual node to start new flows and enquire about the status of the running flows.|
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 26]
```diff
@@ -0,0 +1,43 @@
...
+## Querying Permissions via HTTP RPC
+
+At the moment it is only possible to retrieve a permission if its identifier is known to the caller.
+However, there might be cases when a new role is being setup and existing permissions need to be included. It would be helpful to do a search of the permissions available.
+
+{{< note >}}
+Being unable to do this creates potentially duplicated permissions.
+{{< /note >}}
+
+## Check Permissions When Starting Flows
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:39]:
```suggestion
## Checking Permissions When Starting Flows
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+| Role   | Description |
+| ----------- | ----------- |
+| `UserAdminRole` | Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.       |
+| `VNodeCreatorRole` | Set all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
+| `CordaDeveloperRole` | Users can use VNode reset and Vnode status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
+| `FlowExecutorRole`|  Once the vNode is created, a new CLI command will permit creation of such role for a given vNode to start new flows and enquire about the status of the running flows.|
+
+## Querying Permissions via HTTP RPC
+
+At the moment it is only possible to retrieve a permission if its identifier is known to the caller.
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:32]:
```suggestion
At the moment, it is only possible to retrieve a permission if its identifier is known to the caller.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+However, there might be cases when a new role is being setup and existing permissions need to be included. It would be helpful to do a search of the permissions available.
```
Comment (AnnaK-R3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:33]:
```suggestion
However, there might be cases when a new role is being set up and existing permissions need to be included. It would be helpful to do a search of the permissions available.
```

**Review by vkolomeyko — DISMISSED** [review]

**Review by vkolomeyko — DISMISSED** [review]
LGTM

**Review by AnnaK-R3 — APPROVED** [review]

**Review by yarivus — APPROVED** [review]
LGTM

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+{{< note >}}
+These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
+{{< /note >}}
+
+| Role   | Description |
+| ----------- | ----------- |
+| `UserAdminRole` | Creates roles, permissions, and controls all associations between the user, roles, and permissions. This role is created at cluster bootstrap by the admin user. This role is created via REST call enabling to have complete audit trail of the operation performed.       |
+| `VNodeCreatorRole` | Sets all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
+| `CordaDeveloperRole` | Users can use virtual node reset and virtual node status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
+| `FlowExecutorRole`|  Once the virtual node is created, a new CLI command will permit creation of such role for a given virtual node to start new flows and enquire about the status of the running flows.|
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:28]:
```suggestion
| `FlowExecutorRole`|  Once the virtual node is created, a new CLI command permits the creation of such a role for a given virtual node to start new flows and enquire about the status of the running flows.|
```
I don't understand this. What is the "new" command? Are you giving permission to the vnode or to a user to run flows on the vnode?

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+
+## Querying Permissions via HTTP RPC
+
+At the moment, it is only possible to retrieve a permission if its identifier is known to the caller.
+However, there might be cases when a new role is being set up and existing permissions need to be included. It would be helpful to do a search of the permissions available.
+
+{{< note >}}
+Being unable to do this creates potentially duplicated permissions.
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:36]: unable to do what?

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 31]
```diff
@@ -0,0 +1,43 @@
...
+| `VNodeCreatorRole` | Sets all the necessary permissions to create a virtual node, including CPI upload and CPI update. Create this role at cluster bootstrapping time.|
+| `CordaDeveloperRole` | Users can use virtual node reset and virtual node status update. This role should be provisioned at cluster boostrap time and should re-use previously created permissions for other roles.|
+| `FlowExecutorRole`|  Once the virtual node is created, a new CLI command will permit creation of such role for a given virtual node to start new flows and enquire about the status of the running flows.|
+
+## Querying Permissions via HTTP RPC
+
+At the moment, it is only possible to retrieve a permission if its identifier is known to the caller.
+However, there might be cases when a new role is being set up and existing permissions need to be included. It would be helpful to do a search of the permissions available.
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:33]: "It would be helpful to do a search of the permissions available."

Helpful for what? How is the search done?

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 32]
```diff
@@ -0,0 +1,43 @@
...
+
+{{< note >}}
+Being unable to do this creates potentially duplicated permissions.
+{{< /note >}}
+
+## Checking Permissions When Starting Flows
+
+Currently, Corda checks if a user can execute startFlow RPC operations. No checks are made to whether the user can start a particular flow.
+
+These checks should be performed against the RBAC sub-system even before passing start request to FlowWorker.
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:43]:
```suggestion
These checks should be performed against the RBAC sub-system before passing the start request to a FlowWorker.
```

File: content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md [hunk 33]
```diff
@@ -0,0 +1,43 @@
...
+
+* A dedicated role which can create users, roles, permissions and drive all the associations between them.
+* A dedicated role with a set of all the necessary permissions to create a virtual node (including CPI upload).
+* A dedicated role which allows flows to run on this virtual node.
+
+{{< note >}}
+These roles and permissions enable certain common operations. The creation of users and associating such users to these roles should be done manually by an admin user.
+{{< /note >}}
+
+| Role   | Description |
```
Comment (nadinequinr3) [inline content/en/platform/corda/5.0-beta/corda-cli-reference/cli-commands/rpc-permission-templates.md:23]: The descriptions aren't clear. If I understand correctly, each is a role that grants some permissions to a user? So the descriptions should all start with something like:
"Allows a user to... "
or
"Grants the necessary permissions to..."

**Review by yarivus — DISMISSED** [review]
LGTM

**Review by nadinequinr3 — APPROVED** [review]

**Review by yarivus — COMMENTED** [review]
Thx @Niamh25
